### PR TITLE
Save generated ops in files for shareTree fuzz tests

### PR DESCRIPTION
**Pull request by robertobe-ms**
Save generated ops for shareTree fuzz tests in the directory './fuzz-tests-saved-ops' that is relative to the directory where the tests are executed.

### CODE REVIEW

File: packages/test/stochastic-test-utils/src/performActions.ts [hunk 1]
```diff
@@ -219,20 +226,27 @@ export function performFuzzActions<
 
 	for (let operation = generator(state); operation !== done; operation = generator(state)) {
 		operations.push(operation);
-		if (saveInfo !== undefined && operations.length === saveInfo.saveAt) {
-			writeFileSync(saveInfo.filepath, JSON.stringify(operations));
+		if (saveInfo?.saveWhenFinish === false && operations.length === saveInfo.saveAt) {
```
Comment (yann-achard-MS) [inline packages/test/stochastic-test-utils/src/performActions.ts:229]: `saveInfo?.saveWhenFinish === false` will evaluate to `false` when `saveWhenFinish` is undefined.
This will prevent `saveAt` from working unless `saveWhenFinish` is explicitly set to `false`.

File: packages/test/stochastic-test-utils/src/performActions.ts [hunk 2]
```diff
@@ -111,21 +111,28 @@ export async function performFuzzActionsAsync<
 		operation = await generator(state)
 	) {
 		operations.push(operation);
-		if (saveInfo !== undefined && operations.length === saveInfo.saveAt) {
-			await fs.writeFile(saveInfo.filepath, JSON.stringify(operations));
+		if (saveInfo?.saveWhenFinish === false && operations.length === saveInfo.saveAt) {
```
Comment (yann-achard-MS) [inline packages/test/stochastic-test-utils/src/performActions.ts:114]: Are you making these two options mutually exclusive because they will write to the same file?
Someone stepping through a run whose `SaveInfo` has both the `saveAt` and `saveWhenFinish` turned on, would likely be confused by the fact that the operations are not saved after stepping beyond the iteration that `saveAt` was meant to save.

I think saving twice is less confusing, ideally to different paths.
I see you already save to a modified path for the failure case, maybe it would be best to do that for the success case too.

File: packages/test/stochastic-test-utils/src/types.ts [hunk 3]
```diff
@@ -60,6 +60,7 @@ export type AsyncWeights<TOp, TState> = [
 export interface SaveInfo {
 	saveAt?: number;
 	saveOnFailure: boolean;
+	saveWhenFinish: boolean;
 	/** Filepath to dump the history file. Containing folder must already be created. */
 	filepath: string;
```
Comment (yann-achard-MS) [inline packages/test/stochastic-test-utils/src/types.ts:65]: @Abe27342 - What do you think of changing the contract for this field to not include the file extension?
This would allow the performActions.ts code to customize the full path (and automatically add a `.json` extension).
Would making such a change break a bunch of automated tooling that we already have?

File: packages/test/stochastic-test-utils/src/performActions.ts [hunk 4]
```diff
@@ -219,20 +226,27 @@ export function performFuzzActions<
 
 	for (let operation = generator(state); operation !== done; operation = generator(state)) {
 		operations.push(operation);
-		if (saveInfo !== undefined && operations.length === saveInfo.saveAt) {
-			writeFileSync(saveInfo.filepath, JSON.stringify(operations));
+		if (saveInfo?.saveWhenFinish === false && operations.length === saveInfo.saveAt) {
+			writeFileSync(saveInfo.filepath, JSON.stringify(operations, undefined, 4));
```
Comment (yann-achard-MS) [inline packages/test/stochastic-test-utils/src/performActions.ts:230]: @Abe27342 - what do you think of the added indentation here? Is the added file size likely to be an issue? If so, we could make it configurable.

File: packages/dds/tree/src/test/shared-tree/sharedTreeFuzzTests.ts [hunk 5]
```diff
@@ -227,8 +230,15 @@ function runBatch(
 	for (let i = 0; i < batchSize; i++) {
 		const runSeed = seed + i;
 		const generatorFactory = () => take(opsPerRun, opGenerator());
+		const saveInfo: SaveInfo = {
+			saveOnFailure: true,
+			saveWhenFinish: true,
```
[the remark this answers is not in the document]
Comment (robertobe-ms) [reply packages/dds/tree/src/test/shared-tree/sharedTreeFuzzTests.ts:235]: Right now the tests are marked as skip, but I'll take your sugestion.

File: packages/test/stochastic-test-utils/src/types.ts (hunk 3 above)
Comment (Abe27342) [reply packages/test/stochastic-test-utils/src/types.ts:65]: I didn't put a large amount of design thought into this feature, so no objections to changing it. That said, if we are going to do filepath manipulation from within this package, we should be exposing APIs to directly replay a fuzz test from a filepath, as otherwise the user of performFuzzActions will need to infer the same filepath manipulations to locate their file.

Making the JSON extension added by the package is arguably more OK since it's the only format that it can actually write in, but it's still a bit awkward.

Is there a problem with save info on failure being dumped to the same place as save info on success? I'm not sure this actually matters for most reasonable workflows you want to do with fuzz testing. I think I originally intended `saveAt` to be used to generate documents even on successful cases, but in practice I don't think it actually has much value except for "save after all operations". So perhaps we can replace `saveAt` with `saveOnSuccess`.

There isn't any automated tooling to deal with these replay files. And making breaking changes in the package is OK (it's @fluid-internal, thus should only be used from within this repo). But if you change the filepath logic, I'd expect you need to change other usages of `performFuzzActions` which save filepaths to disk for local debugging.

File: packages/dds/tree/src/test/shared-tree/sharedTreeFuzzTests.ts [hunk 6]
```diff
@@ -227,8 +230,15 @@ function runBatch(
 	for (let i = 0; i < batchSize; i++) {
 		const runSeed = seed + i;
 		const generatorFactory = () => take(opsPerRun, opGenerator());
+		const saveInfo: SaveInfo = {
+			saveOnFailure: false, // Change to true to save failing runs.
+			saveOnSuccess: false, // Change to true to save successful runs.
+			filepath: resolve(
+				`./src/test/shared-tree/fuzz-tests-saved-ops/ops_with_seed_${runSeed}`,
```
Comment (Abe27342) [inline packages/dds/tree/src/test/shared-tree/sharedTreeFuzzTests.ts:237]: i'm not sure resolving with relative path to cwd is safe for all workflows/ways one might run the test command. i've generally preferred using a path relative to `__dirname` (which does get a bit ugly since the file being executed is inside the dist folder).

File: packages/test/stochastic-test-utils/src/types.ts (hunk 3 above)
Comment (yann-achard-MS) [reply packages/test/stochastic-test-utils/src/types.ts:65]: The main drive for the filepath manipulations was that `saveAt` and `saveOnSuccess` can step on each other's output. If you think we can ditch `saveAt` and just offer `saveOnSuccess` then I'd much rather we do that.
The only other motivation is that it's nice to know from the log file whether it's the output of a run that failed, but if we really want that we can add it later by supporting multiple paths.

So, to recap, the path forward on this specific issue becomes:
* Remove the `saveAt` option
* Don't manipulate the path to add things like `_fail` or `_success`
* Don't add a `.json` extension (the caller can do that themselves)

File: packages/test/stochastic-test-utils/src/performActions.ts [hunk 7]
```diff
@@ -99,15 +100,16 @@ export async function performFuzzActionsAsync<
 	const applyOperation: (operation: TOperation) => Promise<TState> = async (op) =>
 		(await reducer(state, op)) ?? state;
 
+	if (saveInfo?.filepath !== undefined) {
+		await fs.mkdir(path.dirname(saveInfo.filepath), { recursive: true });
+	}
```
Comment (yann-achard-MS) [inline packages/test/stochastic-test-utils/src/performActions.ts:105]: This goes against the contract advertised on `SaveInfo.filepath`. It would also result in the folder being created even when no file will be written to it, which not a big problem but can be an annoyance sometimes.

I'm leaning towards either keeping the contract as it was or automatically creating the folder when we actually want to write a file to it (we'd probably put all the folder/file creation code in a pair of sync/async functions for terseness). @Abe27342 do you have any thoughts this?

Comment (Abe27342) [reply packages/test/stochastic-test-utils/src/performActions.ts:105]: i think second option there is best (automatically create the folder when about to write). perf of checking the containing path exists repeatedly for runs with lots of failures isn't really a concern here, and it removes some boilerplate from callers.

File: packages/test/stochastic-test-utils/src/types.ts [hunk 8]
```diff
@@ -68,8 +68,8 @@ export type AsyncWeights<TOp, TState> = [
 ][];
 
 export interface SaveInfo {
-	saveAt?: number;
 	saveOnFailure: boolean;
+	saveOnSuccess?: boolean;
 	/** Filepath to dump the history file. Containing folder must already be created. */
```
Comment (yann-achard-MS) [inline packages/test/stochastic-test-utils/src/types.ts:73]: Don't forget to update this comment. :)

File: packages/test/stochastic-test-utils/src/performActions.ts [hunk 9]
```diff
@@ -99,30 +100,51 @@ export async function performFuzzActionsAsync<
 	const applyOperation: (operation: TOperation) => Promise<TState> = async (op) =>
 		(await reducer(state, op)) ?? state;
 
+	let occurredFirstWrite = false;
```
Comment (yann-achard-MS) [inline packages/test/stochastic-test-utils/src/performActions.ts:103]: The fact that we re-throw the error when there is one means we'll only ever save one file for a given run. That means `occurredFirstWrite` will never be used after being set to `true`.
I recommend just removing the logic associated with it. `mkdir` doesn't overwrite anyway.

**Review by yann-achard-MS — APPROVED** [review]
